cpp/src/arrow/util/value_parsing.cc

```diff
@@ -79,5 +86,46 @@ bool StringToFloatConverter::StringToFloat(const char* s, size_t length, double*
   return true;
 }
 
+// ----------------------------------------------------------------------
+// strptime-like parsing
+
+class StrptimeTimestampParser : public TimestampParser {
+ public:
+  explicit StrptimeTimestampParser(std::string format) : format_(std::move(format)) {}
+
+  bool operator()(const char* s, size_t length, TimeUnit::type out_unit,
+                  value_type* out) const override {
+    arrow_vendored::date::sys_seconds time_point;
+    std::istringstream stream(std::string(s, length));
+    if (stream >> arrow_vendored::date::parse(format_, time_point)) {
```

Comment:
Shouldn't you also check that the stream is exhausted? Otherwise trailing data will be ignored.

Reply:
This won't work anyway because HowardHinnant/date doesn't completely work with gcc < 7